jerry-core/vm/vm-stack.c

```diff
@@ -323,62 +366,71 @@ vm_stack_find_finally (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
         exception = jcontext_take_exception ();
       }
 
-      ecma_value_t iterator = stack_top_p[-3];
-      ecma_value_t result = ecma_op_get_method_by_magic_id (iterator, LIT_MAGIC_STRING_RETURN);
+      ecma_value_t result;
 
-      if (!ECMA_IS_VALUE_ERROR (result) && !ecma_is_value_undefined (result))
+      if (context_type == VM_CONTEXT_ITERATOR)
       {
-        if (!ecma_is_value_object (result) || !ecma_op_is_callable (result))
-        {
-          ecma_free_value (result);
-          result = ecma_raise_type_error (ECMA_ERR_MSG ("Iterator 'return' is not callable"));
-        }
-        else
-        {
-          ecma_object_t *return_obj_p = ecma_get_object_from_value (result);
-          result = ecma_op_function_call (return_obj_p, iterator, NULL, 0);
-          ecma_deref_object (return_obj_p);
+        result = ecma_op_iterator_close (stack_top_p[-2]);
+      }
+      else
+      {
+        ecma_value_t iterator = stack_top_p[-3];
+        result = ecma_op_get_method_by_magic_id (iterator, LIT_MAGIC_STRING_RETURN);
 
-          if (context_type == VM_CONTEXT_FOR_AWAIT_OF && !ECMA_IS_VALUE_ERROR (result))
+        if (!ECMA_IS_VALUE_ERROR (result) && !ecma_is_value_undefined (result))
+        {
+          if (!ecma_is_value_object (result) || !ecma_op_is_callable (result))
           {
-            ecma_extended_object_t *async_generator_object_p = VM_GET_EXECUTABLE_OBJECT (frame_ctx_p);
-
-            result = ecma_promise_async_await (async_generator_object_p, result);
+            ecma_free_value (result);
+            result = ecma_raise_type_error (ECMA_ERR_MSG ("Iterator 'return' is not callable"));
+          }
+          else
+          {
+            ecma_object_t *return_obj_p = ecma_get_object_from_value (result);
+            result = ecma_op_function_call (return_obj_p, iterator, NULL, 0);
+            ecma_deref_object (return_obj_p);
 
-            if (!ECMA_IS_VALUE_ERROR (result))
+            if (context_type == VM_CONTEXT_FOR_AWAIT_OF && !ECMA_IS_VALUE_ERROR (result))
             {
-              uint16_t extra_flags = (ECMA_EXECUTABLE_OBJECT_DO_AWAIT_OR_YIELD
-                                      | (ECMA_AWAIT_FOR_CLOSE << ECMA_AWAIT_STATE_SHIFT));
-              async_generator_object_p->u.class_prop.extra_info |= extra_flags;
+              ecma_extended_object_t *async_generator_object_p = VM_GET_EXECUTABLE_OBJECT (frame_ctx_p);
 
-              stack_top_p = vm_stack_context_abort (frame_ctx_p, stack_top_p);
+              result = ecma_promise_async_await (async_generator_object_p, result);
```

Comment:
Is this part is coming from somwhere else? Is it a code duplication?

Reply:
Nope, that's an already existing code, just indented by one. That's why it seems so big change.